### PR TITLE
Task manager users and login

### PR DESCRIPTION
**Pull request by ben-doorloop**

### CODE REVIEW

File: .gitignore
Comment (ben-doorloop) [inline .gitignore:1]: Shouldn't delete this, please restore (can do `git revert #commit-hash` on the specific commit)

File: TaskManager/client/client.js [hunk 1]
```diff
@@ -1,7 +1,11 @@
-let tasks = null; //JSON.parse(localStorage.getItem("tasks")) || [];
-const tasks_url = "http://localhost:5000/tasks";
-let completedTasks = JSON.parse(localStorage.getItem("completedTasks")) || [];
+let tasks = [];
+let completedTasks = [];
 let firstRun = true;
+const routes = {
+	tasks: "http://localhost:5000/tasks",
+	finishTask: "http://localhost:5000/finishTask",
+	login: "http://localhost:5000/login",
```
Comment (ben-doorloop) [inline TaskManager/client/client.js:7]: It's best to not explicitly state the domain, so it'll work wherever you run it, locally or online
```suggestion
	tasks: "/tasks",
	finishTask: "/finishTask",
	login: "/login",
```

File: TaskManager/client/login.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	"password",
+	"loginBtn",
+];
+
+keysArray.forEach((key) => {
+	DOM[key] = document.querySelector(`.${key}`);
+});
+
+const routes = {
+    index: "http://localhost:5000/viewtasks",
```
Comment (ben-doorloop) [inline TaskManager/client/login.js:14]: Omit domains

File: TaskManager/client/login.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+const routes = {
+    index: "http://localhost:5000/viewtasks", 
+	tasks: "http://localhost:5000/tasks",
+	finishTask: "http://localhost:5000/finishTask",
+	login: "http://localhost:5000/",
+};
+
+const login = () => {
+    const username = DOM.username.value;
+    const password = DOM.password.value;
```
Comment (ben-doorloop) [inline TaskManager/client/login.js:22]:
```suggestion
    const username = DOM.username.value.trim().toLowerCase();
    const password = DOM.password.value.trim().toLowerCase();

```

File: TaskManager/client/login.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    
+    // POST response in server.js l.86
+    fetch(routes.login, {
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+            username: username,
+            password: password,
```
Comment (ben-doorloop) [inline TaskManager/client/login.js:36]:
```suggestion
            username,
            password,
```
Shorthand writing :)

File: TaskManager/client/login.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        })
+    }).then((res) => {res.json().then((auth) => {
+            console.log(auth);
+            if (auth.success) {
+                window.location.href = routes.index;
+            } else {
+                DOM.loginForm.reset();
+                DOM.username.focus();
+                const error = document.createElement('div');
+                error.id = 'error';
```
Comment (ben-doorloop) [inline TaskManager/client/login.js:46]: Avoid working with element IDs, this code could easily lead to ID duplicates

File: TaskManager/client/signup.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	"password",
+	"signupBtn",
+];
+
+keysArray.forEach((key) => {
+	DOM[key] = document.querySelector(`.${key}`);
+});
+
+const routes = {
+    index: "http://localhost:5000/viewtasks",
```
Comment (ben-doorloop) [inline TaskManager/client/signup.js:14]: Domains

File: TaskManager/client/signup.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    // POST response in server.js l.129
+    fetch(routes.signup, {
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+            username: username,
+            password: password,
```
Comment (ben-doorloop) [inline TaskManager/client/signup.js:36]:
```suggestion
            username,
            password,
```

File: TaskManager/server.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+};
+
+// DB Setup
+const allTasksFilePath = './db/allTasks';
+const getAllTasks = () => JSON.parse(fs.readFileSync(allTasksFilePath, 'utf8'))
+const updateAllTasks = (allTasks) => fs.writeFileSync(allTasksFilePath, JSON.stringify(allTasks));
+const usersFilePath = './db/users';
+const getUsers = () => JSON.parse(fs.readFileSync(usersFilePath, 'utf8'));
+const updateUsers = (users) => fs.writeFileSync(usersFilePath, JSON.stringify(users));
+let users = getUsers();
```
Comment (ben-doorloop) [inline TaskManager/server.js:33]:
```suggestion
const users = getUsers();
```

File: TaskManager/server.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+	const username = req.body.username;
+	const password = req.body.password;
+	const auth = {
+		success: false,
+		role: null,
+		msg: 'Invalid credentials. Please try again.',
+	};
+
+	users.find((user) => {
```
Comment (ben-doorloop) [inline TaskManager/server.js:97]: So, the point of `.find()` is to iterate over an array and _return_ the first element that was true(y)
Here, you're iterating over it just to iterate (you're not doing anything with what it returns) so tahles a `forEach` would've been the same
I would still use `find`, but receive what it returns, ask if it's truey (meaning it found a user) and then return ok, else return not ok

File: TaskManager/server.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+			auth.success = true;
+			auth.role = user.role;
+			auth.msg = `Welcome, ${username}! You are logged in as ${user.role}.`;
+		}
+	});
+	console.log(`Authentication is ${auth.success}`);
+	console.log(auth.msg);
+	res.json(auth);
+
+	// FIX - code below disfunctional, implemented on clientside instead (login.js l.40).
```
Comment (ben-doorloop) [inline TaskManager/server.js:108]:
```suggestion
	// FIX - code below dysfunctional, implemented on clientside instead (login.js l.40).
```
😉

File: TaskManager/server.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+	// Redirect (no fetch request or link in client)
+app.get(routes.index, (req, res) => {
+	res.sendFile(__dirname + '/client/viewtasks.html');
+});
+
+	// Signup POST (fetch request in signup.js l.29)
+app.post(routes.signup, (req, res) => {
+	const username = req.body.username;
+	const password = req.body.password;
```
Comment (ben-doorloop) [inline TaskManager/server.js:131]: Salting! Let's talk about it!!

File: TaskManager/server.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	const user = {
+		username: username,
+		password: password,
```
Comment (ben-doorloop) [inline TaskManager/server.js:134]:
```suggestion
		username,
		password,
```

File: TaskManager/server.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+		role: 'user',
+	};
+	const users = getUsers();
+	users.push(user);
+	updateUsers(users);
```
Comment (ben-doorloop) [inline TaskManager/server.js:139]:
```suggestion
	updateUsers([...users, user]);
```
Another shorthand writing you can employ :) Can read about the [Spread Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)

**Review by ben-doorloop — COMMENTED** [review]
Amazing work amigoman!!! Very excite to be on this end of a PR!! 💪🏻

File: TaskManager/server.js (hunk 10 above)
Comment (Gilad-Bron) [reply TaskManager/server.js:108]: Why doesn't the redirect work on server side?

**Pull request closed without merging**